Add support for multiple inputs to out_wrapper and strict dtype checking

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78616
* #78615
* #78350
* #80219
* #80443
* __->__ #79941
* #80516

When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `@out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

lezcano committed:
Add support for multiple inputs to out_wrapper and strict dtype checking

When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `@out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

mruberry reviewed (approved):
Ship it!

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

Lezcano commented:
@ezyang This PR is failing on a meta test, but I'm failing to track down where is the issue coming from. Now, I think it's coming from some torch_function modification you did. Any guesses?

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -323,9 +323,8 @@ def meta_index_Tensor(self, indices):
     return self.new_empty(before_shape + replacement_shape + after_shape)
 
 
-@torch.library.impl(meta_lib, "addbmm")
```
Doesn't this explain why the test is failing?

Lezcano replied:
That was indeed wrong (corrected) but was not the cause of the error. The thing is that somewhere the `out` parameter (which is a tuple) is being transformed into several kwargs, each with the name of the corresponding output. This is wrong, and I have not been able to track down where it happens, but it seems related to the contextmanager `enable_torch_function_mode(inner)`...

To see the backtrace of the failures see https://github.com/pytorch/pytorch/runs/7062769044?check_suite_focus=true

ezyang replied:
The function in question here is being registered as an ATen op directly in the dispatcher. Theseget invoked by the function: parseIValuesToPyArgsKwargs

```
  // Populate positional arguments
  for (const auto idx : c10::irange(positional_default_start)) {
    PyTuple_SET_ITEM(
        args.ptr(), idx, schemaAwareToPyObject(idx).release().ptr());
  }
      
  // Populate keyword arguments
  for (const auto idx : c10::irange(kwarg_only_start, arguments.size())) {
    // But don't populate default keyword arguments
    if (is_default(idx))
      continue;
    const auto& arg = schema.arguments()[idx];
    kwargs[py::cast(arg.name())] = schemaAwareToPyObject(idx);
  }   
  return std::make_pair(std::move(args), std::move(kwargs));
```

There is no special handling for out arguments. The dispatcher just slaps them exactly as they look in native_functions.yaml

I'm not convinced I want to change this behavior. But that means we will need two different implementations of out handling, one for refs (which are Python visible) and another for the ATen registrations (which will see them unpacked.)

Lezcano replied:
Couldn't this be solved just by packing the `out` parameters back into a tuple in that context manager before feeding them back to the ref?

ezyang replied:
Yes, but the main point is you need two versions of the function, one to be a ref and one to actually register to the dispatcher

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

Lezcano commented on torch/_decomp/__init__.py:
```diff
@@ -37,7 +40,49 @@ def clamp_min(x):
 
-    def decomposition_decorator(f):
+    def decomposition_decorator(f: Callable) -> Callable:
+        sig = inspect.signature(f)
```
@rgommers all this block of code is the juggling with `inspect` and annotations I told you about. given a function that takes a kwargs `out=` which is a tuple of tensors and returns a namedtuple, it wraps it as a function that takes the each of the arguments as a separate tensor.

The dirty parts are:
- Accessing `out_annotations.__origin__` to check that that type is indeed a `Tuple` 
- Accessing `return_annotation._fields` to get the names from the `NamedTuple` from the return.

rgommers replied:
Thanks @Lezcano, interesting. This is more like codegen (new annotation from old annotation) than actual use of type annotations at runtime to do anything functional, so my concern of using annotations at runtime (which is a no no) doesn't really apply. I'm not surprised that there's no clean ways in the language or in `typing`/`typing_extensions` to do this kind of thing.

lezcano committed:
Update on "Add support for multiple inputs to out_wrapper and strict dtype checking"


When a function returns multiple parameters in PyTorch, the `out`
parameter takes a tuple of tensors (see `linalg.svd` for example).
The current implementation in `out_wrapper_multi` modelled this wrong,
as it assumed that it would take a number of different named
parameters.

This PR implements the correct behaviour in `out_wrapper`. As a small
side-effect, we now need to call `out_wrapper()` when the output is
just one tensor.

This PR also implements an additional optional parameter that checks
whether the dtype of the given `out` is exactly the dtype that the meta
function requires. This is the behaviour that we currently have in
PyTorch, and this check is necessary in eager when we call with these
tensors into external libraries.

We also make the functions with several outputs return a namedtuple,
similar to what we do in PyTorch.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -g

ezyang commented:
@pytorchbot merge -f

suo commented:
@pytorchbot revert -m "broke master https://hud.pytorch.org/pytorch/pytorch/commit/dc7066a8f00f60bd055e78100b144f3373b11aba" -c landrace
